### PR TITLE
Datatable sync

### PR DESCRIPTION
**Pull request by nekosaur**
Made the following changes:

- added synced `pagination` object prop, and moved `page`, `rowsPerPage`, `sortBy` and `descending` into it (solves #498)
- v-model is now optional and used for selected array (together with regular prop `selectedKey`) (solves #400)
- incoming data moved to regular `items` prop
- items slot now exposes a boolean `selected` which can be used in `v-checkbox`
- removed `defaultSort` as pagination prop can be used to same effect

This change allows for external control of pagination, sorting and selection.

Selected example (selectedKey defaults to 'id')
```vue
<v-data-table
  v-model="selected"
  :items="items"
  :headers="headers"
  select-all
  >
  <template slot="items" scope="props">
    <td><v-checkbox primary v-model="props.selected"></v-checkbox></td>
    <td>#{{ props.item.id }}</td>
    <td>{{ props.item.first_name }}</td>
    <td>{{ props.item.email }}</td>
    <td>{{ props.item.gender }}</td>
  </template>
</v-data-table>
...
<script>
  export default {
    data () {
      return {
        selected: []
      }
    }
  }
</script>
```

Pagination example (change default sort)
```vue
<v-data-table
  :pagination.sync="pagination"
  :items="items"
  :headers="headers"
  select-all
  >
  <template slot="items" scope="props">
    <td><v-checkbox primary v-model="props.selected"></v-checkbox></td>
    <td>#{{ props.item.id }}</td>
    <td>{{ props.item.first_name }}</td>
    <td>{{ props.item.email }}</td>
    <td>{{ props.item.gender }}</td>
  </template>
</v-data-table>
...
<script>
  export default {
    data () {
      return {
        pagination: {
          sortBy: 'first_name'
        }
      }
    }
  }
</script>
```

### CODE REVIEW

File: src/components/tables/DataTable.js [hunk 1]
```diff
@@ -177,56 +177,59 @@ export default {
...
-      this.$emit('update', {
-        page: this.page,
-        rowsPerPage: this.rowsPerPage,
-        sorting: this.sorting,
-        desc: this.desc
-      })
+    isSelected (item) {
+      return this.selected && this.selected.find(i => i[this.selectedKey] === item[this.selectedKey])
     },
     sort (index) {
```
Comment (johnleider) [inline src/components/tables/DataTable.js]: I think that this method could be cleaned up a bit. The same call is made for every scenario, the only thing that changes is the object that is emitted. Assign the value of that object and only have one this.$emit.

File: src/components/tables/DataTable.js [hunk 2]
```diff
@@ -115,36 +105,51 @@ export default {
       type: Boolean,
       default: false
     },
-    defaultSort: {
+    selected: {
```
Comment (johnleider) [inline src/components/tables/DataTable.js]: I think we should let the selected be controlled under the model and the items be under a prop items, and just sync that. Would be curious to hear others thoughts.

```v-model="yourSelectedRows"```
```:items.sync="tableItems"```

File: src/components/tables/DataTable.js [hunk 3]
```diff
@@ -177,56 +177,59 @@ export default {
...
     }
   },
 
   mounted () {
-    const header = this.headers.find(h => !('sortable' in h) || h.sortable)
-    this.desc = this.defaultSort ? this.defaultSort.desc : this.desc
-    this.sorting = this.defaultSort ? this.defaultSort.field : header.value
+    const firstSortable = this.headers.find(h => !('sortable' in h) || h.sortable)
+
+    let defaultPagination = {
```
Comment (johnleider) [inline src/components/tables/DataTable.js]: should be const as it doesn't change value, could also just put it into the Object.assign arguments.

**Comment by avengerweb** [comment]
Why default sort is part of pagination? Let's locate it in headers maybe?

**Comment by nekosaur** [comment]
@avengerweb It's part of the synced prop so that sorting can be controlled outside of the data table. The end result is that it can also be used to set the default sorted column. Placing an option in the headers as well duplicates functionality which already exists.

File: src/components/tables/DataTable.js [hunk 4]
```diff
@@ -3,7 +3,7 @@ import Body from './mixins/body'
 import Foot from './mixins/foot'
 import Progress from './mixins/progress'
 import { getObjectValueByPath } from '../../util/helpers'
-
+import Vue from 'vue'
```
Comment (johnleider) [inline src/components/tables/DataTable.js]: What's this guy for?

**Comment by johnleider** [comment]
All reviewers have until Sunday to provide feedback.

**Pull request merged**